daemons/gptp/common/wltimestamper.hpp

```diff
@@ -157,7 +212,21 @@ class WirelessTimestamper : public Timestamper {
 
 	void timingMeasurementConfirmCB(LinkLayerAddress addr, WirelessDialog *dialog);
 	void timeMeasurementIndicationCB(LinkLayerAddress addr, WirelessDialog *current, WirelessDialog *previous, uint8_t *buf, size_t buflen);
-};
 
+	uint64_t _32to64(uint64_t curr) {
+		uint64_t ret;
+		rollover_lock->lock();
+		if (curr < last_count && (last_count - curr) >(MAX_NSEC >> 1)) {
+			rollover = rollover + 1;
+		}
+		ret = rollover;
+		ret *= MAX_NSEC;
+		ret += curr;
+		last_count = curr;
+		rollover_lock->unlock();
+		return ret;
+	}
```

Comment:
The idea is that the driver doesn’t do the “right thing” and this is a kludge until it is fixed.  I guess I would ask that since this is not a main-line branch and that it is being released to “spur” development within AVnu that you disregard this for now.

From: Levi Pearson [mailto:notifications@github.com]
Sent: Friday, December 04, 2015 2:46 PM
To: AVnu/Open-AVB
Cc: Hall, Christopher S
Subject: Re: [Open-AVB] Added support for Intel wireless adapters on Windows (#324)


In daemons/gptp/common/wltimestamper.hpp<https://github.com/AVnu/Open-AVB/pull/324#discussion_r46741725>:

>

> +    uint64_t _32to64(uint64_t curr) {

> +           uint64_t ret;

> +           rollover_lock->lock();

> +           if (curr < last_count && (last_count - curr) >(MAX_NSEC >> 1)) {

> +                   rollover = rollover + 1;

> +           }

> +           ret = rollover;

> +           ret *= MAX_NSEC;

> +           ret += curr;

> +           last_count = curr;

> +           rollover_lock->unlock();

> +           return ret;

> +    }

The redefinition of MAX_NSEC definitely made it clearer why it was defined to the value it was. I think this function, uint64_t _32to64(uint64_t curr), could use a better name. Maybe something to do with rollover handling.

As a general question, and not something you necessarily need to resolve: Is it correct that the source of the value you're working with here is expressed in tens of nanoseconds (i.e. each tick is 10ns) and comes from a 32-bit register? Is this scaling factor part of the WiFi spec, or is it something that could differ between WiFi chipsets? It might improve clarity to separate out the rollover management from the scaling-to-ns process, but I haven't thought about it deeply enough to be sure.

—
Reply to this email directly or view it on GitHub<https://github.com/AVnu/Open-AVB/pull/324/files#r46741725>.